runtime/auth/aws-signing-standard/common/src/aws/smithy/kotlin/runtime/auth/awssigning/standard/Canonicalizer.kt

```diff
@@ -0,0 +1,215 @@
+            if (predicate && value != null) params[name] = value
+        }
+
+        param("Host", builder.url.host, !(signViaQueryParams || "Host" in params))
+        param("X-Amz-Algorithm", ALGORITHM_NAME, signViaQueryParams)
+        param("X-Amz-Credential", credentialValue(config, credentials), signViaQueryParams)
+        param("X-Amz-Content-Sha256", hash, addHashHeader)
+        param("X-Amz-Date", config.signingDate.format(TimestampFormat.ISO_8601_CONDENSED))
+        param("X-Amz-Expires", config.expiresAfter?.inWholeSeconds?.toString(), signViaQueryParams)
+        param("X-Amz-Security-Token", sessionToken, !config.omitSessionToken) // Add pre-sig if omitSessionToken=false
+
+        val headers = builder
+            .headers
+            .entries()
+            .filter { includeHeader(it.key, config) }
+            .associateTo(mutableMapOf()) { it.key.lowercase() to it.value }
+            .entries
+            .sortedBy { it.key }
+        val signedHeaders = headers.joinToString(separator = ";") { it.key }
+
+        param("X-Amz-SignedHeaders", signedHeaders, signViaQueryParams)
+
+        val requestString = buildString {
+            appendLine(builder.method.name)
+            appendLine(builder.url.canonicalPath(config))
+            appendLine(builder.url.canonicalQueryParams())
+            headers.map { it.canonicalLine() }.forEach(::appendLine)
+            appendLine() // Yes, add an extra blank line after all the headers
+            appendLine(signedHeaders)
+            append(hash)
+        }
+        logger.debug { "Canonical request:\n$requestString" }
+
+        param("X-Amz-Security-Token", sessionToken, config.omitSessionToken) // Add post-sig if omitSessionToken=true
+
+        return CanonicalRequest(builder, requestString, signedHeaders, hash)
+    }
+
+    /**
+     * Calculate a hash for this [HttpBody]. This method does not support attempting to calculate a hash on a
+     * non-replayable stream.
+     * @return The hash as a hex string
+     */
+    private suspend fun HttpBody.calculateHash(): String = when (this) {
+        is HttpBody.Empty -> BodyHash.EmptyBody.hash!!
+        is HttpBody.Bytes -> bytes().hash(sha256Supplier).encodeToHex()
+        is HttpBody.Streaming -> {
+            require(isReplayable) { "Stream must be replayable to calculate a body hash" }
+            val reader = readFrom()
+            reader.sha256().encodeToHex().also { reset() }
+        }
+    }
+
+    /**
+     * Calculate a hash for this [SdkByteReadChannel].
+     * @return The hash as a byte array
+     */
+    private suspend fun SdkByteReadChannel.sha256(): ByteArray {
+        val hash = sha256Supplier()
+
+        val sink = ByteArray(STREAM_CHUNK_BYTES)
+        while (!isClosedForRead || availableForRead > 0) {
+            val bytesRead = readAvailable(sink)
+            if (bytesRead <= 0) break
+            hash.update(sink.sliceArray(0 until bytesRead))
```

Comment:
No, it [delegates down to `copyOfRange`](https://github.com/JetBrains/kotlin/blob/92d200e093c693b3c06e53a39e0b0973b84c7ec5/libraries/tools/kotlin-stdlib-gen/src/templates/Filtering.kt#L973-L996) which explicitly returns a copy (in case you want to modify it or something). I'll add an optimization here.

Reply:
Actually, I realized there's no reason to make a copy of this array _even in the case_ that `bytesRead < STREAM_CHUNK_BYTES`. It's inefficient always.

A better change is to expand `HashFunction` to allow updating based on a subset of the array rather than requiring the whole array be used. I'll make that change instead.